mongo/src/main/java/com/redhat/lightblue/mongo/crud/Translator.java

```diff
@@ -660,6 +636,75 @@ private FieldTreeNode resolve(FieldTreeNode context, Path field) {
         return node;
     }
 
+    private static class FieldInfo {
+        final Path field;
+        final FieldTreeNode fieldMd;
+
+        FieldInfo(Path field,FieldTreeNode fieldMd) {
+            this.field=field;
+            this.fieldMd=fieldMd;
+        }
+    }
+    
+    private FieldInfo resolveFieldForQuery(FieldTreeNode context,Path contextPath,Path field) {
+        int n=field.numSegments();
+        // Two paths: p: the full thing, s: only the relative field
+        // If the field reference moves to a parent of relative field, we use p
+        // otherwise, we use s
+        MutablePath p=new MutablePath(contextPath);
+        MutablePath s=new MutablePath();
+        FieldTreeNode fieldNode=context;
+        for(int i=0;i<n;i++) {
+            String seg=field.head(i);
+            if(field.isIndex(i)) {
+                p.push(seg);
+                if(s!=null)
+                    s.push(seg);
+                if(context instanceof ArrayNode) {
+                    fieldNode=((ArrayField)fieldNode).getElement();
+                } else {
+                    throw Error.get(ERR_INVALID_FIELD,field.toString());
+                }
+            } else {
+                if(Path.THIS.equals(seg)) {
+                    // Don't push anything
+                    ;
+                } else if(Path.PARENT.equals(seg)) {
+                    p.pop();
+                    if(s!=null)
+                        if(s.isEmpty())
+                            s=null;
+                        else
+                            s.pop();
+                    fieldNode=fieldNode.getParent();
+                } else {
+                    p.push(seg);
+                    if(s!=null)
+                        s.push(seg);
+                    fieldNode=fieldNode.resolve(new Path(seg));
+                }
+            }
+        }
+        if(!contextPath.isEmpty()) {
+            // If we are in a nonempty context (i.e. nested query under elemMatch), and our variable
+            // refers to a field above that context, then mongo language can't handle it, and we need
+            // to convert to JS
+            FieldTreeNode trc=fieldNode.getParent();
+            boolean fieldIsUnderContext=false;
+            while(trc!=null) {
+                if(trc==context) {
+                    fieldIsUnderContext=true;
+                    break;
+                } else {
+                    trc=trc.getParent();
+                }
+            }
+            if(!fieldIsUnderContext)
+                throw new NeedsJS();
```

Comment:
Would it make more sense to determine if fieldIsUnderContext before calculating p and s? In the case that javascript is needed it would be better to fail fast.